api/apikey/service.go

```diff
@@ -0,0 +1,124 @@
+package apikey
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"time"
+
+	"github.com/gorilla/securecookie"
+	"github.com/pkg/errors"
+
+	portainer "github.com/portainer/portainer/api"
+)
+
+var ErrInvalidAPIKey = errors.New("Invalid API key")
+
+type apiKeyService struct {
+	apiKeyRepository portainer.APIKeyRepository
+	userRepository   portainer.UserService
+	cache            *apiKeyCache
+}
+
+func NewAPIKeyService(apiKeyRepository portainer.APIKeyRepository, userRepository portainer.UserService) *apiKeyService {
+	return &apiKeyService{
+		apiKeyRepository: apiKeyRepository,
+		userRepository:   userRepository,
+		cache:            NewAPIKeyCache(),
+	}
+}
+
+// HashRaw computes a hash digest of provided base64 encoded raw API key.
+func (a *apiKeyService) HashRaw(rawKey string) ([]byte, error) {
+	decodedRawAPIKey, err := base64.StdEncoding.DecodeString(rawKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to decode raw API key")
+	}
+	hashDigest := sha256.Sum256([]byte(decodedRawAPIKey))
+	return hashDigest[:], nil
+}
+
+// GenerateApiKey generates a base64 encoded raw API key for a user (for one-time display).
+// The generated API key is stored in the cache and database.
+func (a *apiKeyService) GenerateApiKey(user portainer.User, description string) (string, *portainer.APIKey, error) {
+	rawAPIKey := securecookie.GenerateRandomKey(32)
```

Comment:
may suggest having an own implementation of the `GenerateRandomKey()` instead of pulling the `github.com/gorilla/securecookie` package for its helper function.
The function itself is very straight forward:
```
import "crypto/rand"

func GenerateRandomKey(length int) []byte {
	k := make([]byte, length)
	if _, err := io.ReadFull(rand.Reader, k); err != nil {
		return nil
	}
	return k
}
```

besides, we may want to init the `rand.Seed` when creating an instance of the Service for security purpose.

Reply:
would this not be another moving part we would then need to manage (e.g. create a service with an initial random seed)?

It can be done but i don't see the need; any specific reasons as to why - since the `github.com/gorilla/securecookie` dependency will still exist regardless of whether we use this function?